### ISSUE TITLE
Different training loss behavior between pytorch-lightning and pure pytorch

### ISSUE THREAD

**Issue by dalek-who** [Lightning-AI/lightning#6704]
## 🐛 Bug

<img width="1431" alt="image" src="https://user-images.githubusercontent.com/31960962/112744707-0800ac80-8fd5-11eb-9556-cfd29da8c4b4.png">

You can see three train loss curves here, actually there are **four** curves (in the legend below):
- orange: pl on k80+cu110
- red: pure pytorch on k80+cu110，which is fully overlaped under the orange one, so you can't see it.
- grey: pl on p100+cu101
- blue: pure pytorch on p100+cu101

I seed everything in these two models and set deterministic=True. I check the random init weights and training data order are all the same. The curves of k80 and p100 can't overlap is not surprising, but what confusing me is that:
- two curves on k80 (orange and red) can overlap, yet two curves on p100 (grey and blue) can't overlap
- on p100 (grey and blue), duing the starting several training steps(about 10+ steps), pl and pure pytorch loss value are the same, and two curves **can** overlap, , but then loss value become a little different (about 0.00x), and then more and more different, and two curves can't overlap any more.
- curves on k80  (orange and red) finally convergence, yet on p100 (grey and blue) they stop at 0.5 and can't reduce any more.

Since the strange behavior on p100 can't be reproduced on k80, and my full project is huge, I don't know where the bug could be happen and how to create a smallest demo to report the bug. Any ideas?

#### K80:
* CUDA:
        - GPU:
                - Tesla K80
                - Tesla K80
                - Tesla K80
                - Tesla K80
        - available:         True
        - version:           11.0
* Packages:
        - numpy:             1.19.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.7.1
        - pytorch-lightning: 1.2.4
        - tqdm:              4.58.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.7.2
        - version:           #1 SMP Wed Feb 3 15:06:38 UTC 2021

#### P100:
* CUDA:
        - GPU:
                - Tesla P40
                - Tesla P40
                - Tesla P40
                - Tesla P40
        - available:         True
        - version:           10.1
* Packages:
        - numpy:             1.19.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.4.0
        - pytorch-lightning: 1.2.4
        - tqdm:              4.57.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.6.5
        - version:           #1 SMP Wed Apr 1 17:49:22 CST 2020

**Comment by dalek-who** [comment]
My project is a sentence-pair classification task, use ```Transformers``` implementation of Bert model、adamW、linear warmup scheduler

**Comment by awaelchli** [comment]
I notice that your P100 box uses a different pytorch version. Is it possible to upgrade that to 1.7.1 for a sanity check?

**Comment by dalek-who** [comment]
@awaelchli unfortunately, pytorch on p100 is limited at 1.4.0 due to the driver version

**Comment by awaelchli** [comment]
And downgrading to 1.4 on the other machine?
In my view it is quite critical to have the same versions for this comparison. You have already verified that your Lightning code is equivalent to the PyTorch code on one system. Now, switching to a new system the hardware changes, but also your PyTorch environment. We don't know which change is causing the difference in the training behavior. It would be no surprise to me if different pytorch version causes different results.

**Comment by dalek-who** [comment]
@awaelchli ok，I'll have a try!

**Comment by dalek-who** [comment]
Now I'm sure pytorch version cause this strange behavior, here's my result: 
# k80, torch 1.4.0 vs pl:
<img width="1469" alt="image" src="https://user-images.githubusercontent.com/31960962/112946597-4e3a4500-9168-11eb-9d0c-81e54eadc317.png">

- orange: pytorch
- red: pl

not overlap

# k80, torch 1.7.1  vs pl:
<img width="1454" alt="image" src="https://user-images.githubusercontent.com/31960962/112947611-93ab4200-9169-11eb-8816-f7f8c4a55a98.png">

- red: pytorch
- orange: pl

overlap


# p100, torch 1.4.0 vs pl:
<img width="1443" alt="image" src="https://user-images.githubusercontent.com/31960962/112947694-b178a700-9169-11eb-977e-5f29a9adcfab.png">
- blue: pytorch
- grey: pl

**Comment by dalek-who** [comment]
@awaelchli I wish a warning in the document that there're some different behavior between pl and some torch version. It cause a performance drop in my use case (finally I change some hparams and overcome the drop). At first I was very confused and nearly gave up pl, and it takes a week to debug.

**Comment by awaelchli** [comment]
> I wish a warning in the document that there're some difference behavior between pl and some torch version

Sorry for that. In my experience, for almost all projects I have worked on I saw performance differences if the pytorch version difference was too big. In the end, did you switch from  pyTorch_version: 1.7.1 to 1.4 or the other way around? I'm curious which hparams made a difference?

**Comment by dalek-who** [comment]
Finally I use torch 1.4 (limited by p40 driver), and change lr from 4e-5 to 2e-5. Though torch version cause the difference, I think the drop is caused by the instability of high lr, affected by unknown underlying factors, and accidentally converge to a bad point.

**Comment by awaelchli** [comment]
How do you see it, is there evidence that the unknown underlying factors come from Lightning? Any suggestions how to proceed, or is this resolved for you?

**Comment by dalek-who** [comment]
I still can't find what actually cause the bug on torch 1.4 and finally ignore it, since lower lr can "resolve" performance problem.
